src/library-aux/scala/Any.scala

```diff
@@ -138,13 +138,87 @@ abstract class Any {
    */
   final def isInstanceOf[T0]: Boolean = sys.error("isInstanceOf")
 
-  /** Cast the receiver object to be of type `T0`.
-   *
-   *  Note that the success of a cast at runtime is modulo Scala's erasure semantics.
-   *  Therefore the expression `1.asInstanceOf[String]` will throw a `ClassCastException` at
-   *  runtime, while the expression `List(1).asInstanceOf[List[String]]` will not.
-   *  In the latter example, because the type argument is erased as part of compilation it is
-   *  not possible to check whether the contents of the list are of the requested type.
+  /** Tell the compiler to skip type safety and treat the receiver object to be of type `T0`.
+   *
+   *  This method is intended for cases where the compiler isn't able to know or capable to infer `T0` and neither
+   *  is it possible to check the subtype relation at runtime and hence skipping type safety is the only option.
+   *
+   *  DON'T USE `asInstanceOf` for type checking, it is **not** an alias for something like
+   *  {{{
+   *   this match {
+   *    case x: T0 => x
+   *    case _     => throw ClassCastException("...")
+   *  }}}
+   *  Use pattern matching or [isInstanceOf] for type testing instead and act as desired if the subtype
+   *  relation does not hold (e.g. return a [scala.util.Failure], throw an Exception etc.).
+   *
+   *  On a language level it merely tells the compiler to forget any type information it has about
+   *  the receiver object and treat it as if it had type `T0`. How the compiler manages to transition from the type 
+   *  of the receiver object to `T0` is unspecified and should be considered an implementation detail of the 
+   *  corresponding compiler backend and can vary between target platform and the target version.
+   *  Transitioning to T0 might involve inserting conversions (including boxing/unboxing), casts, default values or
+   *  just no real operation at all.
+   *
+   *  Following some legitimate usages of `asInstanceOf`:
+   *
+   *  - turn a type into an opaque type (idiomatically done in the apply function of the companion
+   *    object of the opaque type)
```

Comment:
If you're talking about Scala opaque type aliases, there is normally no need for casts in this scenario.

Reply:
my bad, I thought it was necessary to tell the compiler, going to remove it